### PR TITLE
[#216]  Removed cursor from api of get_matched*() and get_incoming_links()

### LINKED ISSUE #216: Update get_matched_links(), get_matched_link_types() and get_incoming_links() to remove cursor
Required by https://github.com/singnet/das-atom-db/issues/188

### PR DESCRIPTION
**Pull request by andre-senna**
Resolves #216 
Required in the context of https://github.com/singnet/das-atom-db/issues/188

### CODE REVIEW

File: hyperon_das_atomdb/adapters/ram_only.py [hunk 1]
```diff
@@ -572,31 +577,32 @@ def get_matched_links(
 
         patterns_matched = list(pattern) if (pattern := self.db.patterns.get(pattern_hash)) else []
 
-        if kwargs.get("toplevel_only"):
-            return kwargs.get("cursor"), self._filter_non_toplevel(patterns_matched)
+        if kwargs.get("toplevel_only", False):
+            return self._filter_non_toplevel(patterns_matched)
 
-        return kwargs.get("cursor"), patterns_matched
+        # return [handle for handle, _ in patterns_matched]
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/ram_only.py:583]: should this be deleted?

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 2]
```diff
@@ -1012,55 +1001,32 @@ def _retrieve_hash_targets_value(
...
+            **kwargs: Additional keyword arguments
 
         Returns:
-            MatchedTypesResultT: A tuple containing the cursor position (which can be None if
-            `cursor` is absent in kwargs) and a list of members in the hash targets value.
+            MatchedTargetsListT: List of members in the hash targets value.
         """
         key = _build_redis_key(key_prefix, handle)
-        cursor, members = self._get_redis_members(key, **kwargs)
+        # return list(self._get_redis_members(key, **kwargs))
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:1010]: should this be deleted?

File: hyperon_das_atomdb/database.py [hunk 3]
```diff
@@ -47,15 +47,15 @@
...
 LinkParamsT: TypeAlias = LinkT
 
-IncomingLinksT: TypeAlias = list[str] | list[AtomT]
+HandleListT: TypeAlias = list[HandleT]
 
-MatchedTargetsListT: TypeAlias = list[tuple[str, tuple[str, ...]]]
+IncomingLinksT: TypeAlias = HandleListT | list[AtomT]
 
-HandlesListT: TypeAlias = list[HandleT]
+MatchedTargetsListT: TypeAlias = list[tuple[str, tuple[str, ...]]]
```
Comment (angeloprobst) [inline hyperon_das_atomdb/database.py:54]: maybe?
```suggestion
MatchedTargetsListT: TypeAlias = list[tuple[HandleT, tuple[HandleT, ...]]]
```

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 4]
```diff
@@ -887,7 +877,7 @@ def _apply_index_template(
             key.append(WILDCARD if cursor in target_selected_pos else targets[cursor])
         return _build_redis_key(KeyPrefix.PATTERNS, ExpressionHasher.composite_hash(key))
 
-    def _retrieve_incoming_set(self, handle: str, **kwargs) -> tuple[int | None, list[str]]:
+    def _retrieve_incoming_set(self, handle: str, **kwargs) -> MatchedTargetsListT:
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:880]: docstring is still mentioning `cursor`.

File: hyperon_das_atomdb/adapters/ram_only.py [hunk 5]
```diff
@@ -675,11 +681,7 @@ def add_node(self, node_params: NodeParamsT) -> NodeT | None:
         return node
 
     def add_link(self, link_params: LinkParamsT, toplevel: bool = True) -> LinkT | None:
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/ram_only.py:683]: Return annotation is not optional anymore:
```suggestion
    def add_link(self, link_params: LinkParamsT, toplevel: bool = True) -> LinkT:
```

File: hyperon_das_atomdb/adapters/ram_only.py (hunk 1 above)
Comment (angeloprobst) [reply hyperon_das_atomdb/adapters/ram_only.py:583]: Also, any code related to `UNORDERED_LINK_TYPES` can be deleted.

File: hyperon_das_atomdb/adapters/redis_mongo_db.py [hunk 6]
```diff
@@ -703,45 +679,59 @@ def get_matched_links(
...
+                return []
 
-        link_type_hash = WILDCARD if link_type == WILDCARD else self._get_atom_type_hash(link_type)
+        link_type_hash = (
+            WILDCARD if link_type == WILDCARD else ExpressionHasher.named_type_hash(link_type)
+        )
 
         # NOTE unreachable
         if link_type in UNORDERED_LINK_TYPES:
             target_handles = sorted(target_handles)
```
Comment (angeloprobst) [inline hyperon_das_atomdb/adapters/redis_mongo_db.py:692]: ditto.

**Review by angeloprobst — APPROVED** [review]

File: hyperon_das_atomdb/adapters/ram_only.py (hunk 1 above)
Comment (andre-senna) [reply hyperon_das_atomdb/adapters/ram_only.py:583]: Fixed the former but not the later

File: hyperon_das_atomdb/adapters/redis_mongo_db.py (hunk 2 above)
Comment (andre-senna) [reply hyperon_das_atomdb/adapters/redis_mongo_db.py:1010]: Fixed

File: hyperon_das_atomdb/database.py (hunk 3 above)
Comment (andre-senna) [reply hyperon_das_atomdb/database.py]: Fixed

File: hyperon_das_atomdb/adapters/redis_mongo_db.py (hunk 4 above)
Comment (andre-senna) [reply hyperon_das_atomdb/adapters/redis_mongo_db.py:880]: Fixed

File: hyperon_das_atomdb/adapters/ram_only.py (hunk 5 above)
Comment (andre-senna) [reply hyperon_das_atomdb/adapters/ram_only.py:682]: Fixed

File: hyperon_das_atomdb/adapters/redis_mongo_db.py (hunk 6 above)
Comment (andre-senna) [reply hyperon_das_atomdb/adapters/redis_mongo_db.py:692]: Not fixed. I just created a card for this: https://github.com/singnet/das-atom-db/issues/218

**Pull request merged**